File: app/src/main/java/com/admin/theapp/dagger/AppModule.java
package com.admin.theapp.dagger;

import android.arch.lifecycle.ViewModelProvider;
import android.content.Context;

import com.admin.theapp.BuildConfig;
import com.admin.theapp.HotelsApp;
import com.admin.theapp.base.ViewModelFactory;
import com.admin.theapp.logger.AppLogger;
import com.admin.theapp.logger.ReleaseLogger;
import com.theapp.tools.Logger;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module(subcomponents = ViewModelSubComponent.class)
abstract class AppModule {

    @Provides
    @Singleton
    static Context provideContext(HotelsApp application) {
        return application.getApplicationContext();
    }

    @Provides
    @Singleton
    static ViewModelProvider.Factory provideViewModelFactory(ViewModelSubComponent.Builder viewModelSubComponent) {
        return new ViewModelFactory(viewModelSubComponent.build());
    }

    @Provides
    @Singleton
    static Logger provideLogger() {
        return BuildConfig.DEBUG ?
                new AppLogger() :
                new ReleaseLogger();
    }
}

File: app/src/main/java/com/admin/theapp/dagger/AppModule.kt
package com.admin.theapp.dagger

import android.arch.lifecycle.ViewModelProvider
import com.admin.theapp.BuildConfig
import com.admin.theapp.HotelsApp
import com.admin.theapp.base.ViewModelFactory
import com.admin.theapp.logger.AppLogger
import com.admin.theapp.logger.ReleaseLogger
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module(subcomponents = [(ViewModelSubComponent::class)])
object AppModule {

    @Provides
    @JvmStatic
    @Singleton
    fun provideContext(application: HotelsApp) = application.applicationContext

    @Provides
    @JvmStatic
    @Singleton
    fun provideViewModelFactory(viewModelSubComponent: ViewModelSubComponent.Builder): ViewModelProvider.Factory =
        ViewModelFactory(viewModelSubComponent.build())

    @Provides
    @JvmStatic
    @Singleton
    fun provideLogger() = if (BuildConfig.DEBUG) AppLogger() else ReleaseLogger()
}
